### PR TITLE
Bugfix, renaming arguments

### PR DESCRIPTION
**Pull request by code-lukas**
Hi Marios,

I implemented the changes we discussed and added a handful of basic tests. I searched the other files for usages of `dim`, but please have a look on your own to see if I missed anything.

Let me know if there is anything you are missing.

Best,

Lukas

### CODE REVIEW

File: hnne/projector.py [hunk 1]
```diff
@@ -150,7 +177,8 @@ def fit(
             self,
             X: np.ndarray,
             y: np.ndarray = None,
-            dim: int = 2,
+            dim: Union[Optional[int], None] = None,
```
Comment (koulakis) [inline hnne/projector.py:180]: The type can be simplified to `Optional[int]`: https://docs.python.org/3/library/typing.html#typing.Optional

File: hnne/projector.py [hunk 2]
```diff
@@ -150,7 +177,8 @@ def fit(
...
+            override_dim: Union[Optional[int], None] = None,
```
Comment (koulakis) [inline hnne/projector.py:181]: Same here.

File: hnne/projector.py [hunk 3]
```diff
@@ -165,7 +193,10 @@ def fit(
         y: array, shape (n_samples, )
             Ignored.
 
-        dim: int (default 2)
+        dim: Union[Optional[int], None] (default None) (deprecated)
```
Comment (koulakis) [inline hnne/projector.py:196]: And here, only optional.

File: hnne/projector.py [hunk 4]
```diff
@@ -165,7 +193,10 @@ def fit(
...
+            Argument used to overwrite the original dimension of the target space of the projection.
+
+        override_dim: Union[Optional[int], None] (default None)
```
Comment (koulakis) [inline hnne/projector.py:199]: Same here.

File: hnne/projector.py [hunk 5]
```diff
@@ -86,13 +94,32 @@ class HNNE(BaseEstimator):
     """
     def __init__(
             self,
-            dim: int = 2,
+            dim: Union[Optional[int], None] = None,
```
Comment (koulakis) [inline hnne/projector.py:97]: Here I would keep the type as int and set its default to 2. It is nice to have this as a default as it is the most common case.

File: hnne/projector.py [hunk 6]
```diff
@@ -86,13 +94,32 @@ class HNNE(BaseEstimator):
...
+            n_components: Union[int, None] = None,
```
Comment (koulakis) [inline hnne/projector.py:98]: Same here, `int = 2`

File: hnne/projector.py [hunk 7]
```diff
@@ -86,13 +94,32 @@ class HNNE(BaseEstimator):
...
             metric: str = 'cosine',
             radius: float = 0.4,
             ann_threshold: int = 40000,
             preliminary_embedding: str = 'pca'
     ):
-        self.dim = dim
+        if (dim is None) and (n_components is None):
```
Comment (koulakis) [inline hnne/projector.py:104]: This first check is not needed if the `n_components` defaults to 2.

File: hnne/projector.py [hunk 8]
```diff
@@ -86,13 +94,32 @@ class HNNE(BaseEstimator):
...
+            raise ValueError('Either `dim` or `n_components` must be specified.')
+
+        if (dim is not None) and (n_components is not None):
```
Comment (koulakis) [inline hnne/projector.py:107]: To reproduce the old behavior with minimal confusion, we could override the `n_components` with the value of `dim` if `n_components == 2`, else raise the error (as it is clear that the user set both values).

File: tests/test_deprecations.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
+from hnne import HNNE
```
Comment (koulakis) [inline tests/test_deprecations.py:1]: Nice, tests are very welcome :smiley:

**Comment by koulakis** [comment]
Hi @code-lukas. Thanks a lot for preparing the MR, I added some comments :-) The overall code looks good.

I would also suggest to run all tests as some are failing e.g. with `pytest .`. They should be fixed though if you make some of the suggested changes (on the default value of `n_components`).

**Comment by code-lukas** [comment]
I made some adjustments to accommodate your changes @koulakis. All tests ran green in my pre-commit, same with the latest commit. I added some tests for the default functionality if no arguments are specified. The only 'shortcut' I had to take was to set `n_components = 2 in the constructor and not immediately in the method signature as I would otherwise be unable to differentiate between a user-specified `n_components = 2` and the default.

**Comment by koulakis** [comment]
That was fast! I like the solution of defaulting to 2 only if both values are none. 

One last nitpick, could you please change in line 99 of the projector code the type of `n_components` from `Union[int, None]` to `Optional[int]`? Then we can also completely remove the import of the `Union` type as it is not used anywhere else.

**Comment by code-lukas** [comment]
Slipped right by me, thanks for the heads-up! Should be fixed now

**Pull request merged**